### PR TITLE
Support prefixed deploymentConfig name

### LINKED ISSUE #4182: oc deploy doesn't take "dc/dc-name" format
`oc deploy` doesn't take "dc/dc-name" format, making it painful to pipe commands.

@ironcladlou

### PR DESCRIPTION
**Pull request by ironcladlou**
Support prefixed deploymentConfig name references to allow command
chaining. Refactor the code to use client interfaces. Update all tests
to use fake clients.

Fixes #4182

### CODE REVIEW

**Comment by ironcladlou** [comment]
cc @kargakis

File: pkg/cmd/cli/cmd/deploy.go [hunk 1]
```diff
@@ -158,60 +166,33 @@ func (o DeployOptions) Validate(args []string) error {
...
 
-func (o DeployOptions) RunDeploy() error {
-	config, err := o.osClient.DeploymentConfigs(o.namespace).Get(o.deploymentConfigName)
+func (o *DeployOptions) RunDeploy() error {
+	r := o.builder.
+		NamespaceParam(o.namespace).
+		ResourceTypeOrNameArgs(false, o.deploymentConfigName).
+		SingleResourceType().
+		Do()
+	if r.Err() != nil {
```
Comment (kargakis) [inline pkg/cmd/cli/cmd/deploy.go]: No need for this, `r.Object` is checking it too.

File: pkg/cmd/cli/cmd/deploy.go [hunk 2]
```diff
@@ -224,31 +205,16 @@ func (o DeployOptions) RunDeploy() error {
...
-	deployment, err := c.client.GetDeployment(config.Namespace, deploymentName)
+	deployment, err := o.kubeClient.ReplicationControllers(config.Namespace).Get(deploymentName)
 	if err != nil {
-		if !kerrors.IsNotFound(err) {
-			return err
+		if kerrors.IsNotFound(err) {
+			return fmt.Errorf("couldn't find deployment %q", deploymentName)
 		}
+		return err
 	} else {
```
Comment (kargakis) [inline pkg/cmd/cli/cmd/deploy.go]: This else clause is redundant now

File: pkg/cmd/cli/cmd/deploy.go (hunk 1 above)
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Fixed

File: pkg/cmd/cli/cmd/deploy.go (hunk 2 above)
Comment (ironcladlou) [inline pkg/cmd/cli/cmd/deploy.go]: Fixed

**Comment by ironcladlou** [comment]
@kargakis feedback addressed, PTAL

**Comment by kargakis** [comment]
LGTM

Once https://github.com/kubernetes/kubernetes/pull/10979 is in, we should use it.

**Comment by ironcladlou** [comment]
[merge]

**Pull request merged**